File: brdgme-go/category_5/command.go
package category_5

import (
	"errors"
	"strconv"

	"github.com/brdgme/brdgme/brdgme-go/brdgme"
)

type PlayCommand struct {
	Card Card
}

type ChooseCommand struct {
	Row int
}

func (g *Game) Command(
	player int,
	input string,
	players []string,
) (brdgme.CommandResponse, error) {
	parser := g.CommandParser(player)
	if parser == nil {
		return brdgme.CommandResponse{}, errors.New(
			"not expecting any commands at the moment",
		)
	}
	output, err := parser.Parse(input, players)
	if err != nil {
		return brdgme.CommandResponse{}, err
	}
	switch value := output.Value.(type) {
	case ChooseCommand:
		return g.ChooseCommand(player, value.Row, output.Remaining)
	case PlayCommand:
		return g.PlayCommand(player, value.Card, output.Remaining)
	}
	return brdgme.CommandResponse{}, errors.New("inexhaustive command handler")
}

func (g *Game) CommandSpec(player int) *brdgme.Spec {
	parser := g.CommandParser(player)
	if parser != nil {
		spec := parser.ToSpec()
		return &spec
	}
	return nil
}

func (g *Game) ChooseCommand(player, row int, remaining string) (brdgme.CommandResponse, error) {
	logs, err := g.Choose(player, row)
	return brdgme.CommandResponse{
		Logs:      logs,
		CanUndo:   false,
		Remaining: remaining,
	}, err
}

func (g *Game) PlayCommand(player int, card Card, remaining string) (brdgme.CommandResponse, error) {
	logs, err := g.Play(player, card)
	return brdgme.CommandResponse{
		Logs:      logs,
		CanUndo:   false,
		Remaining: remaining,
	}, err
}

func (g *Game) CommandParser(player int) brdgme.Parser {
	parsers := brdgme.OneOf{}
	if g.CanChoose(player) {
		parsers = append(parsers, g.ChooseParser())
	}
	if g.CanPlay(player) {
		parsers = append(parsers, g.PlayParser(player))
	}
	if len(parsers) == 0 {
		return nil
	}
	return parsers
}

func (g *Game) ChooseParser() brdgme.Parser {
	min := 1
	max := len(g.Board)
	return brdgme.Map{
		Parser: brdgme.Chain{
			brdgme.Doc{
				Name:   "choose",
				Desc:   "choose the row to take",
				Parser: brdgme.Token("choose"),
			},
			brdgme.AfterSpace(
				brdgme.Doc{
					Name: "row",
					Desc: "the row to take",
					Parser: brdgme.Int{
						Min: &min,
						Max: &max,
					},
				},
			),
		},
		Func: func(value interface{}) interface{} {
			return ChooseCommand{
				Row: value.([]interface{})[1].(int),
			}
		},
	}
}

func (g *Game) PlayParser(player int) brdgme.Parser {
	cards := []brdgme.EnumValue{}
	for _, c := range g.Hands[player] {
		cards = append(cards, brdgme.EnumValue{
			Name:  strconv.Itoa(int(c)),
			Value: c,
		})
	}
	return brdgme.Map{
		Parser: brdgme.Chain{
			brdgme.Doc{
				Name:   "play",
				Desc:   "play a card",
				Parser: brdgme.Token("play"),
			},
			brdgme.AfterSpace(
				brdgme.Doc{
					Name: "card",
					Desc: "the card to play",
					Parser: brdgme.Enum{
						Values: cards,
						Exact:  true,
					},
				},
			),
		},
		Func: func(value interface{}) interface{} {
			return PlayCommand{
				Card: value.([]interface{})[1].(Card),
			}
		},
	}
}

File: rust/game/category-5-2/src/command.rs
use brdgme_game::command::parser::*;

use crate::Game;

#[derive(Debug, PartialEq, Copy, Clone)]
pub enum Command {
    Play(crate::Card),
    Choose(usize),
}

impl Game {
    pub fn command_parser(&self, player: usize) -> Option<Box<dyn Parser<T = Command> + '_>> {
        if self.is_finished() {
            return None;
        }
        let mut parsers: Vec<Box<dyn Parser<T = Command>>> = vec![];
        if self.can_choose(player) {
            parsers.push(Box::new(self.choose_parser()));
        }
        if self.can_play(player) {
            parsers.push(Box::new(self.play_parser(player)));
        }
        if parsers.is_empty() {
            None
        } else {
            Some(Box::new(OneOf::new(parsers)))
        }
    }

    pub fn choose_parser(&self) -> impl Parser<T = Command> {
        Map::new(
            Chain2::new(
                Doc::name_desc("choose", "choose the row to take", Token::new("choose")),
                AfterSpace::new(Doc::name_desc(
                    "row",
                    "the row to take",
                    Int::bounded(1, crate::ROWS as i32),
                )),
            ),
            |(_, row)| Command::Choose(row as usize),
        )
    }

    pub fn play_parser(&self, player: usize) -> impl Parser<T = Command> {
        let mut cards = self.hands.get(player).cloned().unwrap_or_default();
        cards.sort();
        cards.dedup();
        Map::new(
            Chain2::new(
                Doc::name_desc("play", "play a card", Token::new("play")),
                AfterSpace::new(Doc::name_desc(
                    "card",
                    "the card to play",
                    Enum::exact(cards),
                )),
            ),
            |(_, card)| Command::Play(card),
        )
    }
}
